forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,169 @@
+import forge.card.CardStateName;
+import forge.game.Game;
+import forge.game.ability.AbilityKey;
+import forge.game.ability.AbilityUtils;
+import forge.game.ability.SpellAbilityEffect;
+import forge.game.card.*;
+import forge.game.cost.CostDiscard;
+import forge.game.cost.CostPart;
+import forge.game.cost.CostReveal;
+import forge.game.player.Player;
+import forge.game.player.PlayerController;
+import forge.game.spellability.LandAbility;
+import forge.game.spellability.SpellAbility;
+import forge.game.zone.PlayerZone;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.CardTranslation;
+import forge.util.Lang;
+import forge.util.Localizer;
+
+import java.util.List;
+import java.util.Map;
+
+public class DiscoverEffect extends SpellAbilityEffect {
+
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+        final StringBuilder sb = new StringBuilder();
+        List<Player> players = getTargetPlayers(sa);
+        sb.append(Lang.joinHomogenous(players)).append(" discover").append(players.size() > 1 ? " " : "s " );
+        sb.append(sa.getParamOrDefault("Num", "1")).append(".");
+
+        return sb.toString();
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+
+        // Exile cards from the top of your library until you exile a nonland card with <N> mana value or less.
+        String type = "Card.nonLand+cmcLE" + (sa.getParamOrDefault("Num", "1"));
+
+        for (final Player p : getTargetPlayers(sa)) {
+            if (p == null || !p.isInGame()) continue;
+
+            Card found = null;
+            CardCollection exiled = new CardCollection();
+            CardCollection rest = new CardCollection();
+
+            final PlayerZone library = p.getZone(ZoneType.Library);
+
+            for (int i = 0; i < library.size(); i++) {
+                final Card c = library.get(i);
+                exiled.add(c);
+                if (c.isValid(type, sa.getActivatingPlayer(), host, sa)) {
+                    found = c;
+                    if (sa.hasParam("RememberDiscovered"))
+                        host.addRemembered(c);
+                    break;
+                } else {
+                    rest.add(c);
+                }
+            }
+
+            if (exiled.size() > 0) {
+                game.getAction().reveal(exiled, p, false);
+            }
+
+            changeZone(exiled, ZoneType.Exile, game, sa);
+
+            // Cast it without paying its mana cost or put it into your hand.
+            if (found != null) {
+                String prompt = Localizer.getInstance().getMessage("lblDiscoverChoice",
+                                CardTranslation.getTranslatedName(found.getName()));
+                final boolean play = p.getController().chooseBinary(sa, prompt,
+                        PlayerController.BinaryChoiceType.CastOrHand);
```

Comment:
i don't know if need another Binary Choice or if we just do "Yes OR No" 

@Agetian your opinion?

Reply:
That's fair... could make it work either way.